### ISSUE TITLE
Live Markdown Preview

### ISSUE THREAD

**Issue by thienlnam** [Expensify/App#14872]
## Proposal
https://expensify.slack.com/archives/CC7NECV4L/p1675122898707719

Design Doc: https://docs.google.com/document/d/1rcmXjgwvsM3MtdQnkJlMs6IS-i8SHJd0Vs7B_U0K3UU/edit

**Proposal: Implement a Live Markdown Preview feature in NewDot**
**Context:** Markdown is a text formatting syntax that we use in NewDot to add rich text formatting to messages. It allows users to add basic formatting such as bold, italic, bullet points, headings, [links](https://www.youtube.com/watch?v=dQw4w9WgXcQ), etc. to their messages without having to use HTML or other complex formatting tools.

**Problem:**
As NewDot becomes more widely used for group communication, it’s essential to offer ways to effectively format and structure messages to improve productivity and avoid miscommunication similar to how we do [resyncs](https://expensify.slack.com/archives/CC7NECV4L/p1674083559827169) in WN.
The problem is that without a visual representation of the final message, it can be difficult for users to confirm that their message is correctly formatted and conveys the intended information. Inconsistent or improperly formatted messages can hinder effective communication and decrease productivity, especially for users who are not familiar with markdown.

**Solution:**
To enhance user experience and improve communication, add a live markdown preview feature in the chat application. This will allow users to preview their message format in real-time and make any necessary adjustments before sending the message, resulting in clearer and more easily understood communications.
The live markdown preview feature is already present in Slack, and gives you the assurance that the message you are crafting, will show up exactly like you expect once sent.
The implementation will be done in collaboration with Callstack, utilizing the Design Doc process, and will be executed as a concurrent project. The live markdown preview feature is a front-end focused project and will not impact any internal employees or ongoing newsletters.

## Tasks

- [x] Post **Proposal (full Problem/Solution statement)** in `#expensify-open-source`
- [x] Wait at least one full business day, and until the post has a majority (2/3) of positive reactions (👍)
- [x] Paste Proposal in the space above with a link to the Slack thread
- [x] Email `strategy@expensify.com` and paste in the Proposal
- [x] Fill out the **High-level overview of the problem**, **Timeline**, and **Terminology** sections of the Design Doc
- [x] Email `strategy@expensify.com` (continue the same email chain as before) with the link to your Design Doc
- [x] Host a **pre-design** meeting ([example](https://expensify.slack.com/archives/C01GTK53T8Q/p1665794669539419)) in `#expensify-open-source` to discuss any necessary details in public before filling out the **High-level of proposed solution** section.
- [x] Fill out the **High-level of proposed solution** section
- [x] Email `stategy@expensify.com` again with links to the doc and pre-design conversation in Slack
- [x] Add the `DesignDocReview` label to get the **High-level of proposed solution** section reviewed
- [x] Respond to any questions or concerns and bring up blockers in Slack to get a consensus if necessary
- [x] Confirm that the doc has the minimum necessary number of reviews before proceeding
- [x] Host another pre-design meeting in `#expensify-open-source` to ask for engineering feedback on the technical solution.
- [x] Fill out the **Detailed implementation of the solution** and related sections.
- [x] Re-add the `DesignDocReview` label to this issue
- [ ] Respond to any questions or concerns and bring up blockers in Slack to get consensus if necessary
- [ ] Confirm that the doc has the minimum necessary number of reviews before proceeding
- [ ] Email `strategy@expensify.com` one last time to let them know the Design Doc is moving into the implementation phase
- [ ] Implement the changes
- [ ] Send out a follow up email to `strategy@expensify.com` once everything has been implemented and do a **Project Wrap-Up** retrospective that provides:
    - Summary of what we accomplished with this project
    - What went well?
    - What could we have done better?
    - What did we learn?

**Comment by matkoson** [comment]
Hey @thienlnam 👋 I'd love to take care of this feature 🙂

**Comment by MelvinBot** [comment]
@thienlnam, @matkoson Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by MelvinBot** [comment]
@thienlnam, @matkoson Still overdue 6 days?! Let's take care of this!

**Comment by MelvinBot** [comment]
@thienlnam, @matkoson 6 days overdue. This is scarier than being forced to listen to Vogon poetry!

**Comment by thienlnam** [comment]
@matkoson Is currently working on writing out the high level.  Additionally, please make sure to work inside the design doc so we can keep track of what else needs to be done

**Comment by thienlnam** [comment]
@matkoson Is currently out sick for the week, will be back on Monday

**Comment by MelvinBot** [comment]
@thienlnam, @matkoson Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by MelvinBot** [comment]
@thienlnam, @matkoson Huh... This is 4 days overdue. Who can take care of this?

**Comment by matkoson** [comment]
hey @thienlnam, I've updated the 'High-level overview of the problem, Timeline, and Terminology sections' of the Design Doc.

**Comment by MelvinBot** [comment]
@thienlnam, @matkoson Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by thienlnam** [comment]
@matkoson Has been wrapping up the DD high level, and should be finished with it tomorrow

**Comment by thienlnam** [comment]
@matkoson is working on updating the high level and then working on a high level solution that doesn't require a webview

**Comment by matkoson** [comment]
@thienlnam I've updated the high level, and should have no-webview solution proposal by tommorrow.

**Comment by MelvinBot** [comment]
:wave: Hello Generalist Track Team - you have been assigned to review this **High Level** Design Doc. Please, treat this as a priority. Review [this Stack Overflow](https://stackoverflow.com/c/expensify/questions/5838) for some tips on reviewing a design doc. Once you are done, simply press the `Add "Reviewed Doc" comment` button in the right hand side K2 pannel or follow [these](https://stackoverflow.com/c/expensify/questions/13713) instructions.

---
- `@sonialiap` (Expensifier / Graduate) - https://github.com/Expensify/App/issues/15824
- `@mlouiser` (Expensifier / Graduate) - https://github.com/Expensify/App/issues/15825
- `@mateocole` (Project Manager) - https://github.com/Expensify/App/issues/15826
- `@zsgreenwald` (Project Manager) - https://github.com/Expensify/App/issues/15827
- `@isabelastisser` (Product Manager) - https://github.com/Expensify/App/issues/15828
- `@conorpendergrast` (Product Manager) - https://github.com/Expensify/App/issues/15829
- `@quinthar` (Generalist) - https://github.com/Expensify/App/issues/15830
- `@shawnborton` (Generalist) - https://github.com/Expensify/App/issues/15831
- `@kortneywynn` (Accounting Technical Team) - https://github.com/Expensify/App/issues/15832

**Comment by thienlnam** [comment]
Just sent out the high level design doc! Waiting on reviews now while @matkoson is working on figuring out the technical details

**Comment by mallenexpensify** [comment]
@thienlnam , I put the below on hold pending Live Markdown Preview, if you disagree, please comment on the issue below. 
- https://github.com/Expensify/App/issues/15866

**Comment by quinthar** [comment]
Can you link to the design doc in here?  I'm not seeing it.  Perhaps update the original comment to make it super obvious?

**Comment by MelvinBot** [comment]
@thienlnam, @matkoson Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by kortneywynn** [comment]
I have read and reviewed this Design Doc!

**Comment by MelvinBot** [comment]
@thienlnam, @matkoson Eep! 4 days overdue now. Issues have feelings too...

**Comment by sonialiap** [comment]
I have read and reviewed this Design Doc!

**Comment by MelvinBot** [comment]
@thienlnam, @matkoson 6 days overdue. This is scarier than being forced to listen to Vogon poetry!

**Comment by jliexpensify** [comment]
I have read and reviewed this Design Doc!

**Comment by thienlnam** [comment]
We're working on reviews on the high level, @matkoson is working on finalizing the implementation details so we can move forward with the high level detailed

**Comment by isabelastisser** [comment]
I have read and reviewed this Design Doc!

**Comment by roryabraham** [comment]
I have read and reviewed this Design Doc!

**Comment by conorpendergrast** [comment]
I have read and reviewed this Design Doc!

**Comment by shawnborton** [comment]
I have read and reviewed this Design Doc!

**Comment by mlouiser** [comment]
I have read and reviewed this Design Doc!

**Comment by trjExpensify** [comment]
I have read and reviewed this Design Doc!

**Comment by MelvinBot** [comment]
@thienlnam, @matkoson Huh... This is 4 days overdue. Who can take care of this?

**Comment by thienlnam** [comment]
Ready to get cranking on the detailed - @matkoson any update here?

**Comment by matkoson** [comment]
I'll be updating the issue today and leave another comment once it's done.

**Comment by MelvinBot** [comment]
@thienlnam, @matkoson Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by matkoson** [comment]
👋, I've updated the design doc!

**Comment by matkoson** [comment]
@thienlnam I've updated the detailed section [here](https://docs.google.com/document/d/1rcmXjgwvsM3MtdQnkJlMs6IS-i8SHJd0Vs7B_U0K3UU/edit#heading=h.ns12dr1f35rx) doc. I believe all the comments have been addressed as well 👍

**Comment by MelvinBot** [comment]
:wave: Hello Generalist Track Team - you have been assigned to review this **Detailed** Design Doc. Please, treat this as a priority. Review [this Stack Overflow](https://stackoverflow.com/c/expensify/questions/5838) for some tips on reviewing a design doc. Once you are done, simply press the `Add "Reviewed Doc" comment` button in the right hand side K2 pannel or follow [these](https://stackoverflow.com/c/expensify/questions/13713) instructions.

---
- `@arielgreen` (Expensifier / Graduate) - https://github.com/Expensify/App/issues/17514
- `@caitlinwhite1` (Expensifier / Graduate) - https://github.com/Expensify/App/issues/17515
- `@zsgreenwald` (Project Manager) - https://github.com/Expensify/App/issues/17516
- `@strepanier03` (Project Manager) - https://github.com/Expensify/App/issues/17517
- `@robertjchen` (Product Manager / Generalist + Engineer) - https://github.com/Expensify/App/issues/17518
- `@yuwenmemon` (Product Manager / Generalist + Engineer) - https://github.com/Expensify/App/issues/17519
- `@AndrewGable` (Product Manager / Generalist + Engineer) - https://github.com/Expensify/App/issues/17520
- `@tgolen` (Product Manager / Generalist + Engineer) - https://github.com/Expensify/App/issues/17521
- `@conorpendergrast` (Product Manager / Generalist + Non-Engineer) - https://github.com/Expensify/App/issues/17522
- `@quinthar` (Product Manager / Generalist + Non-Engineer) - https://github.com/Expensify/App/issues/17523
- `@isabelastisser` (Product Manager / Generalist + Non-Engineer) - https://github.com/Expensify/App/issues/17524
- `@flaviadefaria` (Product Manager / Generalist + Non-Engineer) - https://github.com/Expensify/App/issues/17525

**Comment by thienlnam** [comment]
Detailed doc is complete, sending it out for review!

**Comment by tgolen** [comment]
I have read and reviewed this Design Doc!

**Comment by cead22** [comment]
I have read and reviewed this Design Doc!

**Comment by strepanier03** [comment]
I have read and reviewed this Design Doc!

**Comment by MelvinBot** [comment]
@thienlnam, @matkoson Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by arielgreen** [comment]
I have read and reviewed this Design Doc!

**Comment by isabelastisser** [comment]
I have read and reviewed this Design Doc!

**Comment by matkoson** [comment]
I have addressed most of the outstanding comments. However, I think the questions related to potential selection and caret positioning issues warrant a dedicated section with detailed explanations and code snippets. I wasn't able to complete that section today, but I expect to finish it tomorrow. cc @thienlnam

**Comment by puneetlath** [comment]
I have read and reviewed this Design Doc!

**Comment by matkoson** [comment]
Hey 👋  @thienlnam, I have added a new section addressing the issue of caret and selection positioning, including two possible solutions, a proposed course of action, and all the considerations that I could think of. Please let me know if there are any improvements you think I can make to the section - https://docs.google.com/document/d/1rcmXjgwvsM3MtdQnkJlMs6IS-i8SHJd0Vs7B_U0K3UU/edit#bookmark=id.mhsbx2sicvf

**Comment by yuwenmemon** [comment]
I have read and reviewed this Design Doc!

**Comment by roryabraham** [comment]
I have read and reviewed this Design Doc!

**Comment by AndrewGable** [comment]
I have read and reviewed this Design Doc!

**Comment by JmillsExpensify** [comment]
I have read and reviewed this Design Doc!

**Comment by robertjchen** [comment]
I have read and reviewed this Design Doc!

**Comment by matkoson** [comment]
I've addressed all the comments, including those from yesterday. It seems like the next step is to build a POC for the "Fake Caret" approach as well. @thienlnam

**Comment by thienlnam** [comment]
We've hit a bit of a hiccup in the detailed portion, @matkoson is going to start a convo in #expensify-open-source regarding our options

**Comment by quinthar** [comment]
I have read and reviewed this Design Doc!

**Comment by thienlnam** [comment]
Latest convo is happening here: https://expensify.slack.com/archives/C01GTK53T8Q/p1683321621449219

TLDR: Detailed solution may not work well for us and we are looking into alternative solutions

**Comment by trjExpensify** [comment]
I have read and reviewed this Design Doc!

**Comment by thienlnam** [comment]
There was a discussion in slack regarding our approach in the detailed: https://expensify.slack.com/archives/C01GTK53T8Q/p1683321621449219

@matkoson Created a mock-up in the doc and updated the doc with it https://expensify.slack.com/archives/C01GTK53T8Q/p1683969682216399?thread_ts=1683321621.449219&cid=C01GTK53T8Q

Unless you have any blockers could you get started with implementation? @matkoson

**Comment by thienlnam** [comment]
Latest conversation here https://expensify.slack.com/archives/C01GTK53T8Q/p1685668673771149?thread_ts=1683321621.449219&cid=C01GTK53T8Q

**Comment by matkoson** [comment]
POC with the new implementation approach progress can be tracked on this branch: https://github.com/matkoson/App/tree/feat/live-markdown-preview-poc

**Comment by roryabraham** [comment]
@matkoson Can you create a draft PR so it's easier to see / comment on the POC?

**Comment by thienlnam** [comment]
Any update here @matkoson?

**Comment by thienlnam** [comment]
Latest in https://expensify.slack.com/archives/C01GTK53T8Q/p1687279801697449?thread_ts=1683321621.449219&cid=C01GTK53T8Q

**Comment by matkoson** [comment]
Hey 👋  I've updated the POC here https://github.com/Expensify/App/compare/main...matkoson:feat/live-markdown-preview-poc?expand=1. OSS post explaining the new approach coming later today.

**Comment by thienlnam** [comment]
Could you please open a draft PR so it's easier for us to review the changes?

**Comment by matkoson** [comment]
@thienlnam on it from yesterday, will post the link here once it's OK for review(aiming for today)

**Comment by matkoson** [comment]
Sorry for the wait. The ETA is a few hours at most. I will keep you updated and provide the PR link in this conversation. @thienlnam

**Comment by thienlnam** [comment]
Could you let us know what blockers you have running into? The last ETA we received was for 5 days ago from https://github.com/Expensify/App/issues/14872#issuecomment-1614468917

Additionally - it would be great to get the slack post [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1686775680269679?thread_ts=1683321621.449219&cid=C01GTK53T8Q) so we can point people towards the new solution

**Comment by matkoson** [comment]
👋 POC code for the Aztec proposal can be found here 👉- https://github.com/Expensify/App/pull/22464

**Comment by TomaszFrackowiak** [comment]
Thread about the Aztec solution - [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1689239226405849)

**Comment by thienlnam** [comment]
Left some comment in the PR but @matkoson is working on making the prototype build on iOS

As for the thread about the Aztec solution - seems like we have general agreement so looking good to move forward here

**Comment by matkoson** [comment]
@thienlnam I've made significant progress on solving native bindings issues - I expect major POC update to be released today, for all platforms.

**Comment by thienlnam** [comment]
@matkoson Did something happen with the update? Noticed the POC still is broken, and has many conflicts

**Comment by matkoson** [comment]
@thienlnam I've dropped the fix! Please have a look when you get the chance 🙏 🙏 🙏

**Comment by matkoson** [comment]
Hey, @jack. I’m almost done with the conflicts, but I’ll be able to finish on August 16th. I’m out of the office today, and tomorrow is a bank holiday in Poland 😕

**Comment by thienlnam** [comment]
Sounds good, will await an update then!

**Comment by thienlnam** [comment]
Mat has been on sick leave, still awaiting the PR to resolve merge conflicts

**Comment by thienlnam** [comment]
Same, discussing options internally

**Comment by thienlnam** [comment]
We're taking a different direction with this so I'm closing out this old issue and creating a new one here: https://github.com/Expensify/App/issues/27977

Thread for context: https://expensify.slack.com/archives/C01SKUP7QR0/p1693954407135859
